File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/7569-토마토/solution.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

public class Main {
    static FastReader scan = new FastReader();
    static StringBuilder sb = new StringBuilder();

    static int N, M;
    static int[][] a;
    static int[][] dist;
    static int[][] dir = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};
    
    static void input() {
        M = scan.nextInt();
        N = scan.nextInt();
        dist = new int[N][M];
        a = new int[N][M];
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                a[i][j] = scan.nextInt();
            }
        }
    }

    static int bfs() {
        Queue<Integer> Q = new LinkedList<>();
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                dist[i][j] = -1;
                if (a[i][j] == 1) {  // 익은 토마토의 위치가 곧 시작점
                    dist[i][j] = 0;
                    Q.add(i);
                    Q.add(j);
                }
            }
        }
        
        // BFS 과정 시작
        while (!Q.isEmpty()) {
            int x = Q.poll();
            int y = Q.poll();
            for (int k = 0; k < 4; k++) {
                int nx = x + dir[k][0], ny = y + dir[k][1];
                if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;  // 지도를 벗어나는 곳으로 가는가?
                if (dist[nx][ny] != -1) continue;  // 이미 방문한 적이 있는 곳인가?
                if (a[nx][ny] == -1) continue;  // 토마토가 들어있지 않는 칸
                Q.add(nx);
                Q.add(ny);
                dist[nx][ny] = dist[x][y] + 1;
            }
        }
        int ans = 0;
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < M; j++) {
                if (a[i][j] == -1) continue;
                if (dist[i][j] == -1) return -1;
                ans = Math.max(ans, dist[i][j]);
            }
        }
        return ans;
    }

    static void pro() {
        System.out.println(bfs());
    }

    public static void main(String[] args) {
        input();
        pro();
    }


    static class FastReader {
        BufferedReader br;
        StringTokenizer st;

        public FastReader() {
            br = new BufferedReader(new InputStreamReader(System.in));
        }

        public FastReader(String s) throws FileNotFoundException {
            br = new BufferedReader(new FileReader(new File(s)));
        }

        String next() {
            while (st == null || !st.hasMoreElements()) {
                try {
                    st = new StringTokenizer(br.readLine());
                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
            return st.nextToken();
        }

        int nextInt() {
            return Integer.parseInt(next());
        }

        long nextLong() {
            return Long.parseLong(next());
        }

        double nextDouble() {
            return Double.parseDouble(next());
        }

        String nextLine() {
            String str = "";
            try {
                str = br.readLine();
            } catch (IOException e) {
                e.printStackTrace();
            }
            return str;
        }
    }
}

File: 강의 자료/02-알고리즘/09~11-그래프 탐색/문제별 코드/7569-토마토/solution.cpp
#include <iostream>
#include <algorithm>
#include <string>
typedef long long int ll;
using namespace std;

#define NM 1005

int N, M, a[NM][NM], dist[NM][NM];
int dir[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

void input() {
    cin >> M >> N;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            cin >> a[i][j];
        }
    }
}

#include <queue>
int bfs() {
    queue<int> Q;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            dist[i][j] = -1;
            if (a[i][j] == 1) {
                dist[i][j] = 0;
                Q.push(i); Q.push(j);
            }
        }
    }
    while (!Q.empty()) {
        int i = Q.front(); Q.pop();
        int j = Q.front(); Q.pop();
        for (int k = 0; k < 4; k++) {
            int ni = i + dir[k][0];
            int nj = j + dir[k][1];
            if (ni < 1 || nj < 1 || ni > N || nj > M) continue;
            if (dist[ni][nj] != -1) continue;
            if (a[ni][nj] == -1) continue;
            dist[ni][nj] = dist[i][j] + 1;
            Q.push(ni); Q.push(nj);
        }
    }

    int ans = 0;
    for (int i = 1; i <= N; i++) {
        for (int j = 1; j <= M; j++) {
            if (a[i][j] == -1) continue;
            if (dist[i][j] == -1) return -1;
            ans = max(ans, dist[i][j]);
        }
    }
    return ans;
}
void pro() {
    cout << bfs();
}


int main() {
    ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
    input();
    pro();
    return 0;
}